source/common/ssl/context_config_impl.h

```diff
@@ -51,9 +59,25 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
   unsigned minProtocolVersion() const override { return min_protocol_version_; };
   unsigned maxProtocolVersion() const override { return max_protocol_version_; };
 
+  bool isReady() const override {
+    // either tls_certficate_provider_ is nullptr or
+    // tls_certficate_provider_->secret() is NOT nullptr.
+    return !tls_certficate_provider_ || tls_certficate_provider_->secret();
+  }
+
+  void setUpdateCallback(Secret::SecretCallbacks& callback) override {
+    if (tls_certficate_provider_.get() != nullptr && secret_callback_ != nullptr) {
+      tls_certficate_provider_->removeUpdateCallback(*secret_callback_);
+    }
+    secret_callback_ = &callback;
+    if (tls_certficate_provider_.get() != nullptr) {
```

Comment:
only save the callback when provider is not null.

You can do

if (secret_callback_) {
    if (secret_callback_) {
           secret_callback_->remove
   }
  secret_callback_ = &callback;
  secret_callback_->add
}

Reply:
Done.